### PR TITLE
[FLINK-22001] Forward exceptions thrown during initialization to the user

### PR DESCRIPTION
**Pull request by rmetzger**
## What is the purpose of the change

- Exceptions during the initialization (thrown by the JobMaster) were not forwarded to the user. This is now guarded by the JobMasterITCase.testJobManagerInitializationExceptionsAreForwardedToTheUser().
- The initialization of the JobMaster (including the scheduler initialization, which runs the source coordinators and other user code) was executed in the JobManager main thread, potentially blocking the JobManager in case of blocking calls in the JobMaster init. 

**Open TODOs**: (due to the amount of changes, and the time spend on the issue already, I'd like to get some initial feedback on the change)
- The problem that the JobMaster initialization was executed in the main thread executor was not, and is still not guarded by a test.
- The newly introduced `JobMasterITCase.testJobManagerInitializationExceptionsAreForwardedToTheUser()` is not nice: It relies on an implementation detail.

## Brief change log

- Introduced `JobManagerStatusListener` to JobManagerRunner
- Reworked the `DispatcherJob` to track the initialization based on the signals from the `JobManagerStatusListener`.
- Introduced a `DispatcherJobStatus` class with tracks the status along with the JobManagerRunner future (they are updated together, thus we track them in one class)
- Adjusted the DispatcherTests because initialization completion is now dependent on getting leadership.

## Verifying this change

Added some tests, adjusted some.

### CODE REVIEW

**Comment by tillrohrmann** [comment]
For the review of the PR it would be great to also capture somewhere the description and analysis of the problem. I know we talked about it, but I already forgot the details ;-)

**Comment by tillrohrmann** [comment]
The problem seems to be the following: Since we create the `JobMasterService` lazily it can happen that the `DispacherJob` is in the initialized state (`JobManagerRunner` being created) but the `JobMasterService` is not running/has not been created. If now the client polls the `DispatcherJob.requestJobStatus()`, the system will ask the `JobManagerRunner.getJobMasterGateway().requestJob()`. The `JobManagerRunner.getJobMasterGateway` might not be completed. 

If now the `JobMasterService` creation fails, then the `JobManagerRunnerImpl` will complete the `resultFuture` which leads to the shut down of the `DispatcherJob` and then also the `JobManagerRunnerImpl`. Due to this shut down, the system will complete the `leaderGatewayFuture` exceptionally which causes the initial `DispatcherJob.requestJobStatus` to fail with `org.apache.flink.util.FlinkException: JobMaster has been shut down.`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobManagerRunnerImpl.java [hunk 1]
```diff
@@ -367,27 +381,39 @@ private void startJobMasterServiceSafely(UUID leaderSessionId) {
...
                             initializationTimestamp);
 
+            // Execute listener notification asynchronously in the main thread executor to make sure
+            // this "grant leadership" operation is properly completed before the next operation is
+            // started. With the current implementation DispatcherJob might call closeAsync()
+            // leading to concurrent access under the "lock".
+            FutureUtils.assertNoException(
+                    CompletableFuture.runAsync(
+                            () -> jobManagerStatusListener.onJobManagerStarted(this),
+                            mainThreadExecutor));
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobManagerRunnerImpl.java:391]: I think there is design problem here. Somehow the `JobManagerRunner` needs to know that some calls have to be executed by the `mainThreadExecutor`. This looks like implementation details which are leaked from the `DispatcherJob` into the `JobManagerRunnerImpl`.

**Review by tillrohrmann — CHANGES_REQUESTED** [review]
Thanks for creating this PR @rmetzger. I fear that this PR is not going in the right direction, tbh.

I think the underlying problem is that you try to mirror the state of the `JobManagerRunner` in the `DispatcherJob` which overly couples these components and possibly not really solves the underlying problem.

If you think about why we introduced the `DispatcherJob`, then it was to handle the asynchronous creation of the `JobManagerRunner`. Now the `JobManagerRunner` can be created synchronously and the asynchronous generation of the `JobMasterService` has been moved to the `JobManagerRunner` implementation. Hence, the `DispatcherJob` is not really doing anything anymore because the `JobManagerRunner` is always created. 

However, what is not created is the `JobMasterService`. Hence, I would suggest to move the `DispatcherJob` logic which manages the job accesses when the job is not yet running into the `JobManagerRunnerImpl`. By doing these accesses under the same lock as the leadership operations, we can ensure that we have a consistent view of the state of the job.

What we probably have to change in the `JobManagerRunnerImpl`, however, is that we don't create the `JobMasterService` under the same lock. Otherwise we risk that the `Dispatcher's` main thread might be blocked while we create the `JobMasterService`.

**Comment by rmetzger** [comment]
Thanks a lot for your review. Next time, I'll write the issue analysis into the Jira (as I usually do).

I agree with your observation that we should remove `DispatcherJob` and move the logic from there into `JobManagerRunnerImpl`. I will rewrite the PR.

**Comment by rmetzger** [comment]
As discussed offline, this is a rough first draft of the revised JobManagerRunnerImpl, pretty much untested.
I'll now start adding tests.

**Comment by rmetzger** [comment]
I've now pushed a version of the change that should have all tests passing, and have a pretty solid test coverage. Looking forward to your feedback @tillrohrmann !

**Review by tillrohrmann — COMMENTED** [review]
Thanks for updating this PR @rmetzger. I think it goes in a very good direction. Moving the responsibility of the `DispatcherJob` into the `JobManagerRunnerImpl` is a good decision imo because it allows us to have a consistent view of the actual state.

I had a couple of comments concerning the current implementation I think that some cases and the semantics of some fields are not fully defined yet. I also think that it could help to make the different states of the `JobManagerRunnerImpl` more explicit and to define how the component should behave in which situation. I haven't managed to create a draft for such a state machine yet but that's what I will continue tomorrow morning with.

**Comment by rmetzger** [comment]
Thanks a lot for your review! Your comments are very helpful!

I thought about representing the states in the Runner more nicely as well, as there are quite a few fields that are only set in certain conditions (cancellation during initialization), and many methods distinguish between two cases (initialized vs initializing). I briefly tried setting up State classes, but it felt that I had to carry a lot of stuff around. Not sure if the pattern with the context used in Adaptive Scheduler isn't a bit of overkill here.
I also realize that some states are still not well defined (initialization error) in the current implementation.

**Comment by rmetzger** [comment]
Closing in favor of https://github.com/apache/flink/pull/15715.

**Pull request closed without merging**